README
===========================================================================

				  Libdfp
	 The "Decimal Floating Point C Library" Developer's Guide
		    for the GNU/Linux OS and GLIBC 2.10
		      Contributed by IBM Corporation
		Copyright (C) 2009 Free Software Foundation

===========================================================================
NOTE:Eight space tabs are the optimum editor setting for reading this file.
===========================================================================

	       Author(s) :  Ryan S. Arnold <rsa@us.ibm.com>
		       Date Created: July 14, 2009
		       Last Changed: August 13, 2009

---------------------------------------------------------------------------
Table of Contents:

	1.  Introduction
	  1.1. ISO/IEC TR 24732
	  1.2. IEEE 754-2008 (DPD & BID Encodings)
	  1.3. Backends (libdecnumber & libbid)
	2.  Availability
	3.  Compliance With ISO/IEC TR 24732
	  3.1 __STDC_WANT_DEC_FP__
	  3.2 scanf
	4.  Dependencies
	  4.1  GNU/Linux OS
	  4.2  GLIBC Minimum Version
	  4.3  GLIBC Headers
	  4.4  libdecnumber
	  4.5  GCC With --enable-decimal-float Support
	  4.6  Autoconf & GNU Make
	5.  Configuration
	  5.1  Configure Switches
	6.  Source Tree Layout
	7.  Make Rules
	8.  Precision (TODO)
	9.  Testing (make check) (TODO)

	A.  History
	B.  Acknowledgements

---------------------------------------------------------------------------
1.  Introduction

The "Decimal Floating Point C Library" is an implementation of ISO/IEC
Technical report  "ISO/IEC TR 24732" which describes the C-Language library
routines necessary to provide the C library runtime support for decimal
floating point data types introduced in IEEE 754-2008, namely _Decimal32,
_Decimal64, and _Decimal128.

---------------------------------------------------------------------------
1.1. ISO/IEC TR 24732

The latest description of ISO/IEC TR 24732 at the time of this writing can
be found here:

	http://www.open-std.org/JTC1/SC22/wg14/www/docs/n1312.pdf

A rationale can be found here:

	http://www.open-std.org/JTC1/SC22/wg14/www/docs/n1242.pdf

---------------------------------------------------------------------------
1.2. IEEE754-2008 (DPD & BID Encodings)

IEEE 754-2008 defines two different encodings for the decimal floating
point data types.  These are DPD and BID.

	DPD (Densely Packed Decimal) - IBM sponsored encoding (implemented
				       in hardware).

	BID (Binary Integer Decimal) - Intel sponsored encoding.

Both encodings can be investigated in the draft IEEE754r:

	http://754r.ucbtest.org/drafts/archive/2006-10-04.pdf

---------------------------------------------------------------------------
1.3. Backends (libdecnumber & libbid)

Libdfp can make use of two backend libraries for software emulation.  These
are "libdecnumber" and "libbid".

Libdecnumber was written by Mike Cowlishaw (IBM) and a form of the library
was donated to the FSF where it was subsumed into GCC.  The Free Software
Foundation assigned GCC as the owner of libdecnumber and now any project that
needs to include libdecnumber statically needs to pull it from the GCC source
trees.  All patches to libdecnumber should be sent to GCC.

This library's copy of libdecnumber should be periodically synced with
upstream GCC's version.  This syncing is the responsibility of the Libdfp
maintainer.

Libbid was written by <TODO> (Intel).

---------------------------------------------------------------------------
2. Availability

Libdfp attempts to provide an encoding agnostic API to users based
upon the _Decimal32, _Decimal64, and _Decimal128 data-types.

Libdfp is available to be configured in the following combinations:

Arch		HW|SW	Backend		Encoding
-------------------------------------------------
ppc970  	SW	libdecnumber	DPD
POWER4		SW	libdecnumber	DPD
POWER		SW	libdecnumber	DPD
POWER6		HW	libdecnumber	DPD
POWER7		HW	libdecnumber	DPD
S/390 z9-109	SW	libdecnumber	DPD
S/390 z9-ec	HW	libdecnumber	DPD
S/390 z10	HW	libdecnumber	DPD
x86		SW	libdecnumber	DPD | BID
x86_64		SW	libbid		BID

The selection of a backend and/or an encoding is made at configuration time
by the distro maintainer or person building libdfp using configuration
flags described in the section entitled "Configuration".

---------------------------------------------------------------------------
3. Compliance With ISO/IEC TR 24732

This section covers issues related to compliance with ISO/IEC TR 24732.

---------------------------------------------------------------------------
3.1 __STDC_WANT_DEC_FP__

The standard ISO/IEC TR 24732 indicates that programs that wish to use
Decimal Floating Point should define the following macro:

	__STDC_WANT_DEC_FP__

There is no set value for this macro.

Simply passing -D__STDC_WANT_DEC_FP__ on compilation, or defining it in your
program should suffice:

#define __STDC_WANT_DEC_FP__

This macro is REQUIRED when including dfp/math.h dfp/fenv.h, etc to pick
up the DFP function prototypes and data types defined in these headers.

---------------------------------------------------------------------------
3.2 scanf

Libdfp does not, and will not comply with the TR 24732 requirement for the
addition of scanf in support of decimal floating point data types.  The
strtod[32|64|128] functions can be used for the same purpose without the
implications of scanf.

---------------------------------------------------------------------------
3.3 printf

Libdfp supports the addition of the printf format codes indicated by TR
24732.  GLIBC proper owns the printf implementation.  Libdfp utilizes the
printf-hooks mechanism provided by GLIBC to register support of the DFP
format codes.

The fully functional printf-hooks mechanism was debuted in GLIBC 2.10.

---------------------------------------------------------------------------
4. Dependencies

---------------------------------------------------------------------------
4.1 GNU/Linux OS

Libdfp is only enabled to work on the GNU/Linux OS.

---------------------------------------------------------------------------
4.2 GLIBC Minimum Version

Libdfp version 1.0.0.0 relies upon a minimum GLIBC 2.10 for printf-hooks
support.

---------------------------------------------------------------------------
4.3 GLIBC Headers

Libdfp must be built against the GLIBC 2.10 headers due to the printf-hooks
mechanism being described there.  Due to a problem with GLIBC's "make
install-headers" not installing all of the necessary headers (such as
feature.h) the developer building libdfp must:

	- `configure' GLIBC for the target
	- `make'
	- `make install_root="<some-path>" install'.
	- tar the headers out of "<some-path>/include".
	- unpack the archive close to the libdfp sources, e.g.
		/home/$USER/stage_libdfp/glibc-2.10-includes/

The resultant headers should be pointed to when Libdfp `configure' is run
(as described in the section on Configuration).

---------------------------------------------------------------------------
4.4 libdecnumber

The libdecnumber library has it's own configure and Makefile fragments.
This is so that it can be built from within any source tree simply by
passing the necessary configure flags.

In reality it's a bit more complicated.

The libdecnumber library requires the --enable-decimal-float flag when
being configured.

This pushes this requirement upon the top level Libdfp 'configure'.   Sadly this
is necessary because Autoconf doesn't allow passing flags to
AC_CONFIG_SUBDIRS.  It only passes the --with and --enable flags of the
parent configure fragment to the child, therefore the parent needs to have
all the flags used by the child.

We used the --enable-decimal-float flag to indicate which encoding to use,
DPD or BID.  The following are valid:

--enable-decimal-float=dpd
--enable-decimal-float=bid
--enable-decimal-float=yes {uses the host default encoding}

Specifying --enable-decimal-float=no simply defaults to the host default
encoding as well.

Some of the files in the Libdfp top directory are symlinked to the actual
files in the ./scripts directory.  This is because libdecnumber lives in
the GCC source tree and GCC doesn't use AC_CONFIG_AUX_DIR([scripts]) but
Libdfp DOES.  This is required so that libdecnumber/configure.ac ->
configure works properly.

---------------------------------------------------------------------------
4.5 GCC With --enable-decimal-float Support

There's a dependency on a version of GCC which supports Decimal Floating
Point.  Use the following to determine if your compiler supports it:

	gcc -v 2>&1 | grep "\-\-enable\-decimal\-float"

If decimal floating point support is not available in your compiler the
libdfp configure stage will fail with a warning.

---------------------------------------------------------------------------
4.6 Autoconf & GNU Make

This Library uses Autoconf to generate GNU Makefiles from Makefile.in
templates and NOT Automake because Automake is incapable of selecting .S or
.c source files from an override directory.  Automake is poorly suited for
system libraries which require flexibility based on assembler code
overrides and submachine architecture overrides.  Automake demands that the
target file list be statically defined.

If you need to re-run autoreconf -i you need to use Autoconf version 2.59
because libdecnumber explicitly requires it.  Here's a hint on how to use a
particular autoconf version:

PATH=/usr/local/autoconf-2.59/bin/:$PATH autoreconf -i

This of course requires that you've installed Autoconf version 2.59 with
--prefix=/usr/local/autoconf-2.59/

Some of the Makefile machinery was inspired by the GLIBC Makefiles,
beautifully written by Roland McGrath since this library was originally a
GLIBC add-on.

---------------------------------------------------------------------------
5. Configuration

Standalone libdfp requires that your build directory be separate from
the source directory.  Invoke the source directorie's 'configure' script
from within the build directory, e.g.

/home/$USER/stage_libdfp/build/$ ../libdfp/configure

---------------------------------------------------------------------------
5.1 Configure Switches

Libdfp's default configure should be 64-bit.  In order to build 32-bit you can
pass the following switches:

	--build=powerpc-linux-gnu --host=powerpc-linux-gnu

The following switches define the DFP backend.

--with-backend={libdecnumber|libbid}

	[OPTIONAL]

	Defaults to host default.  Libbid is not valid on PowerPC or s390
	hardware.

--enable-decimal-float={yes|dpd|bid}

	[REQUIRED if --with-backend=libdecnumber]

	If --with-backend=libbid then anything other than 'bid' or 'yes'
	is invalid.

	'bid' is not valid on PowerPC or s390 hardware.

Valid Combinations:

	--with-backend=libdecnumber --enable-decimal-float=dpd
	--with-backend=libdecnumber --enable-decimal-float=bid
	--with-backend=libbid --enable-decimal-float=bid

--with-glibc-headers=/home/$USER/stage_libdfp/glibc-2.10-headers

	[REQUIRED if your toolchain has GLIBC version less than 2.10]

--with-glibc-build=/home/$USER/stage_glibc/build/glibc64_power6/

	[OPTIONAL]

	If you want to run make check but your system GLIBC isn't version 2.10
	you need to pass the location of a temporary GLIBC build-dir (not
	install dir) where that GLIBC meets or exceeds version 2.10.  Not
	passing this switch tells the make check system to simply use the
	system GLIBC.

	WARNING: Make sure you link to a GLIBC build that supports the same
	configure options as how you're configuring libdfp, e.g. if you're
	configuring for 'power6' 64-bit, make sure that the GLIBC build you
	link against is a 64-bit, power6 tuned build.

--with-cpu={power5|power6|<etc>}

	[OPTIONAL]

	Use this if you want the compiler to optimize the generated code
	for a particular processor.  This also causes the libdfp build
	infrastructure to choose architecture specific overrides from the
	sysdeps tree.

Here are some valid configurations of Libdfp:

General PowerPC (soft-dfp) using the libdecnumber backend and dpd encoding:

CC=/opt/at05/bin/gcc \
/home/ryanarn/eglibc/eglibc/libdfp/trunk/configure \
	--with-backend=libdecnumber --enable-decimal-float=dpd \
	--with-glibc-headers=/home/ryanarn/ppc64-tc-utils/trunk/libdfp/include-2.10/ \
	--with-glibc-build=/home/ryanarn/glibc-2.11/build/glibc64_power5 \
	2>&1 | tee _configure64_power5

PowerPC POWER6 (hard-dfp) using libdecnumber backend and dpd encoding:
We only need the --with-cpu=power6 flag for power6 support and
hardware-dfp:

CC=/opt/at05/bin/gcc \
/home/ryanarn/eglibc/eglibc/libdfp/trunk/configure \
	--with-backend=libdecnumber --enable-decimal-float=dpd \
	--with-glibc-headers=/home/ryanarn/ppc64-tc-utils/trunk/libdfp/include-2.10/ \
	--with-glibc-build=/home/ryanarn/glibc-2.11/build/glibc64_power6 \
	--with-cpu=power6 2>&1 | tee _configure64_power6

PowerPC POWER7 (hard-dfp) using libdecnumber backend and dpd encoding:
We only need the --with-cpu=power7 flag for power7 support and
hardware-dfp:

CC=/opt/at05/bin/gcc \
/home/ryanarn/eglibc/eglibc/libdfp/trunk/configure \
	--with-backend=libdecnumber --enable-decimal-float=dpd \
	--with-glibc-headers=/home/ryanarn/ppc64-tc-utils/trunk/libdfp/include-2.10/ \
	--with-glibc-build=/home/ryanarn/glibc-2.11/build/glibc64_power7 \
	--with-cpu=power7 2>&1 | tee _configure64_power7

To build 32-bit use the following (notice the --build=powerpc-linux-gnu
--host=powerpc-linux-gnu switches):

CC=/opt/at05/bin/gcc \
/home/ryanarn/eglibc/eglibc/libdfp/trunk/configure \
	--with-backend=libdecnumber --enable-decimal-float=dpd \
	--with-glibc-headers=/home/ryanarn/ppc64-tc-utils/trunk/libdfp/include-2.10/ \
	--build=powerpc-linux-gnu --host=powerpc-linux-gnu \
	--with-glibc-build=/home/ryanarn/glibc-2.11/build/glibc32_power5 \
	2>&1 | tee _configure32_power5

Or 32-bit POWER6:

CC=/opt/at05/bin/gcc \
/home/ryanarn/eglibc/eglibc/libdfp/trunk/configure \
	--with-backend=libdecnumber --enable-decimal-float=dpd \
	--with-glibc-headers=/home/ryanarn/ppc64-tc-utils/trunk/libdfp/include-2.10/ \
	--build=powerpc-linux-gnu --host=powerpc-linux-gnu \
	--with-glibc-build=/home/ryanarn/glibc-2.11/build/glibc32_power6 \
	--with-cpu=power6 2>&1 | tee _configure32_power6

Or 32-bit POWER7:

CC=/opt/at05/bin/gcc \
/home/ryanarn/eglibc/eglibc/libdfp/trunk/configure \
	--with-backend=libdecnumber --enable-decimal-float=dpd \
	--with-glibc-headers=/home/ryanarn/ppc64-tc-utils/trunk/libdfp/include-2.10/ \
	--build=powerpc-linux-gnu --host=powerpc-linux-gnu \
	--with-glibc-build=/home/ryanarn/glibc-2.11/build/glibc32_power7 \
	--with-cpu=power7 2>&1 | tee _configure32_power7

---------------------------------------------------------------------------
6. Source Tree Layout

The Libdfp configure script is responsible for building the search order that
is used by the Make system for selecting files out of the source tree.

The configure script automatically knows how to search for
	base_machine (e.g. powerpc)
	machine (e.g. powerpc32)
	submachine (e.g. power6, power7)

It also recognizes the following special directories when they show up
following any of the previous *machine directories:
	fpu
	nofpu
	dfpu
	bits

It hard-searches for the following sysdep directories based upon configuration
settings:
	bid
	dpd
	soft-dfp

And it knows how to find the common files stored in the convenience
directories:

	ieee754r
	base-math

The top of the search list is the commond libdfp internal header files:

	${srcdir}/include.

Anything in these directories should be ABSOLUTELY COMMON since you can't get
a header earlier than in the ${srcdir}/include directory.  If you need to
override something in ${srcdir}/include then you may need to move the headers
out of this directory and into the sysdeps tree.

Header files that are used when building libdfp, but also provided by libdfp
and installed in the prefix directory are in the following directory:

	${srcdir}/dfp

The sysdeps tree is searched and relevant directory depth takes
precedence.

${srcdir}/include [internal libdfp headers]
${srcdir}/libdecnumber/dpd [libdecnumber headers]
${srcdir}/libdecnumber/bid [libdecnumber headers]
${srcdir}/libbid/? [libbid headers]
[MACHINE/SUBMACHINE]
  ${srcdir}/sysdeps/powerpc/powerpc64/power7/fpu [platform directories]
  ${srcdir}/sysdeps/powerpc/powerpc64/power7/
  ${srcdir}/sysdeps/powerpc/powerpc64/power6/fpu [platform directories]
  ${srcdir}/sysdeps/powerpc/powerpc64/power6/
  ${srcdir}/sysdeps/powerpc/powerpc64/
  ${srcdir}/sysdeps/powerpc/powerpc32/power6/fpu [platform directories]
  ${srcdir}/sysdeps/powerpc/powerpc32/power6/
  ${srcdir}/sysdeps/powerpc/powerpc32/
[BASE_MACHINE/[FPU|NOFPU|BITS]]
  ${srcdir}/sysdeps/powerpc/fpu
  ${srcdir}/sysdeps/powerpc/fpu/bits
  ${srcdir}/sysdeps/powerpc/nofpu
  ${srcdir}/sysdeps/powerpc/nofpu/bits
[BASE_MACHINE]
  ${srcdir}/sysdeps/powerpc/
  ${srcdir}/sysdeps/i386/
[ENCODING]
  ${srcdir}/sysdeps/dpd [dpd specific implementations]
  ${srcdir}/sysdeps/bid [bid specific implementations]
[SOFT-DFP ENCODING OVERRIDES[DPD|BID]]
  ${srcdir}/sysdeps/soft-dfp/dpd [soft-dfp dpd arithmetic, conversion, and comparison fns]
  ${srcdir}/sysdeps/soft-dfp/bid [soft-dfp bid arithmetic, conversion, and comparison fns]
${srcdir}/sysdeps/soft-dfp [soft-dfp arithmetic, conversion, and comparison fns]
${srcdir}/dfp [headers files that will be installed.  These follow the
      	 sysdep overrides so that sysdeps override directories show up
      	 before the install headers in #include_next <header>.
${srcdir}/ieee754r [ISO C Draft TR functions]
${srcdir}/base-math [Arithmetic, Conversion, and Comparison Routines.]
${srcdir}/[common libdfp framework code & common headers for the backends.]

Out of Tree Sources
${srcdir}/libdecnumber [third-party GCC sources]
${srcdir}/libbid       [third-party libBID sources]

---------------------------------------------------------------------------
7. Make Rules

make [all (default)]

	DEPENDENCIES:

		GLIBC headers: The location of the companion GLIBC 2.10 (or
		greater) headers as passed to configure using the
		--with-glibc-headers switch.

make check

	DEPENDENCIES:

		GLIBC build: The location of a companion GLIBC 2.10 (or
		greater) build (prior to make install) that is to be linked
		against during the make check run.  This is required because
		the printf-hooks feature that is necessary for printing
		_Decimal[32|64|128] numbers is only in GLIBC 2.10 and later.
		This isn't necessary if the system GLIBC is version 2.10 or
		later.  The path to the build was passed to configure using
		the --with-glibc-build switch.

make install [install_root=<path>]

	[install_root] (Optional) : Install to <path>/$prefix.  This is used
	by libdfp developers and distro builders so that they can build libdfp
	and install it to an alternate location in preparation for packaging.

make install-headers [install_root=<path>]

	[install_root] (Optional) : Install libdfp headers into
	<path>/$prefix/include/dfp.  This is used by application or library
	developers whose projects depend on libdfp who don't want to install
	libdfp proper or may not have permission to do so.

make clean

[TODO] make distclean

---------------------------------------------------------------------------
Appendices
---------------------------------------------------------------------------
A. History

Libdfp was designed by Ryan S. Arnold (IBM), Janis Johnson (IBM), and Steven
J. Munroe (IBM) in 2006.

Libdfp was originally written as a GLIBC add-on by Ryan S. Arnold
<rsa@us.ibm.com> and Joseph Kerian <jkerian@us.ibm.com> in 2006 and
contributed to the Free Software Foundation by IBM in 2007.

Janis Johnson <janis187@us.ibm.com> provided the initial versions of the
arithmetic, conversion, and comparison routines.

The libdecnumber library was originaly contributed to GCC by IBM and written
by Mike Cowlishaw.  It's inclusion into GCC was negotiated in 2005/2006 and it
was pulled into Libdfp as a utility library shortly thereafter.

GLIBC proper declined to include Libdfp upstream on the grounds that
ratification of the DFP technical report had not taken place so...

Libdfp was ported to EGLIBC by Pete Eberlein <eberlein@us.ibm.com> in 2007.

System 390 support was added by Andreas Krebbel <krebbel@de.ibm.com>
starting in 2008.

Ryan S. Arnold (with the help of Andreas Krebbel) ported Libdfp into a
stand-alone library in 2009.

---------------------------------------------------------------------------
B.  Acknowledgements

Thanks to Ulrich Drepper and the GLIBC folks since we used much of the GLIBC
macro black magic, math routines, string-to-float conversions, and printf_fp
code as a basis for the corresponding decimal floating point forms.

Thanks to Roland McGrath for his wild Makefile foo in GLIBC.  Ryan used a lot
of that as inspiration for Libdfp's Makefiles.

Thanks to everyone at IBM (past and present) who worked on Libdfp.

Thanks to Jeff Bailey for helping Ryan grok Autoconf and GNU Make.  His
examples on SUFFIXES rules were invaluable in getting the stand-alone Libdfp
make files into a functional form.

Thanks to EGLIBC (eglibc.org) for hosting Libdfp when it was a GLIBC add-on
and for hosting the stand alone version as well.

Thanks to Michael Matz (SuSE) for help getting the initial packaging for
Libdfp straightened out.



NEWS
Decimal Floating Point C Library - history of user-visible changes.
Copyright (C) 2014-2015 Free Software Foundation, Inc.
See the end for copyright conditions.

Version 1.0.10

* Added BID support for x86_64 and x86.

* Fixes to avoid external PLT calls to internal supplied symbols.

* Add optimized fabsd[32|64|128], and copysignd[32|64|128] for PowerPC
  with DFP hardware support.

----------------------------------------------------------------------

Version 1.0.9

* Support for powerpc64le has beed added.

* Various fixes for LE build.

* Support for configure with --with-cpu=power8 has been added.

* Fix for isinfd[32|64|128]: it now follow GLIBC convention and returns
  1 for plus infinity and -1 for minus infinity.

* Fix finited result to return a positive argument for any normal
  number.

* New <math.h> macro named issignaling to check for a signaling NaN (sNaN).
  It is based on draft TS 18661.

* New llogbd[32|64|128] implementation to return the logb result as
  long int.

* New roundevend[32|64|128] implementation that rounds its argument to the
  nearest integer value in floating-point format, rounding halfway cases
  to even (that is, to the nearest value whose least significant bit 0),
  regardless of the current rounding direction. 
  It is based on draft 18661-2.

* Add optimized ceild[32|64|128], truncd[32|64|128], floord[32|64|128],
  nearbyintd[32|64|128], roundd[32|64|128], and roundevend[32|64|128]
  for PowerPC with DFP hardware support.

* New quantumd[32|64|128] implementation that compute the quantum of a
  finite argument.
  It is based on draft 18661-2.

* Add optimized quantumd[32|64|128] and samequantumd[32|64|128]
  implementation for PowerPC with DFP hardware support.

----------------------------------------------------------------------
Copyright information:

Copyright (C) 2007-2014 Free Software Foundation, Inc.

he Decimal Floating Point C Library is free software; you can
redistribute it and/or modify it under the terms of the GNU Lesser
General Public License version 2.1.

The Decimal Floating Point C Library is distributed in the hope that
it will be useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
the GNU Lesser General Public License version 2.1 for more details.

You should have received a copy of the GNU Lesser General Public
License version 2.1 along with the Decimal Floating Point C Library;
if not, write to the Free Software Foundation, Inc., 51 Franklin
Street, Fifth Floor, Boston, MA 02110-1301 USA.

Please see libdfp/COPYING.txt for more information.


dfp/math.h
/* math.h for libdfp and redirect to system math.h.  This also replaces the
   system math.h classification macros.

   Copyright (C) 2006 IBM Corporation.
   Copyright (C) 2009-2020 Free Software Foundation, Inc.

   This file is part of the Decimal Floating Point C Library.

   Author(s): Ryan S. Arnold <rsa@us.ibm.com>

   The Decimal Floating Point C Library is free software; you can
   redistribute it and/or modify it under the terms of the GNU Lesser
   General Public License version 2.1.

   The Decimal Floating Point C Library is distributed in the hope that
   it will be useful, but WITHOUT ANY WARRANTY; without even the implied
   warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
   the GNU Lesser General Public License version 2.1 for more details.

   You should have received a copy of the GNU Lesser General Public
   License version 2.1 along with the Decimal Floating Point C Library;
   if not, write to the Free Software Foundation, Inc., 51 Franklin
   Street, Fifth Floor, Boston, MA 02110-1301 USA.

   Please see libdfp/COPYING.txt for more information.  */

#ifndef _DFP_MATH_H
#define _DFP_MATH_H 1

#ifdef _MATH_H
# error "dfp/<math.h> should be included before the system math.h."
#endif

/* Include the system math.h first so that we can undef and redefine the
 * classification functions to select for _Decimal[32|64|128] types.  */
#include_next <math.h>

/* Include this after the #include_next <math.h> so that if the user has
 * -I/<path>/include/dfp/ on their compilation line but doesn't define
 *  __STDC_WANT_DEC_FPP that they still get the system math.h.  */
#ifdef __STDC_WANT_DEC_FP__

#include <stdbool.h>

/* math.h provides prototypes for the classification macros below. */
#include <math.h>

#define DEC_INFINITY	__builtin_infd32()
#define DEC_NAN		__builtin_nand32("")
#define HUGE_VAL_D32	__builtin_infd32()
#define HUGE_VAL_D64	__builtin_infd64()
#define HUGE_VAL_D128	__builtin_infd128()

/* Some useful constants for DFP support (with the DL specifier).  Proper
 * truncation to DD and DF will be handled by GCC.  */
#define M_Edl		2.7182818284590452353602874713526625DL  /* e */
#define M_LOG2Edl	1.4426950408889634073599246810018921DL  /* log_2 e */
#define M_LOG10Edl	0.4342944819032518276511289189166051DL  /* log_10 e */
#define M_LN2dl	0.6931471805599453094172321214581766DL  /* log_e 2 */
#define M_LN10dl	2.3025850929940456840179914546843642DL  /* log_e 10 */
#define M_PIdl	3.1415926535897932384626433832795029DL  /* pi */
#define M_PI_2dl	1.5707963267948966192313216916397514DL  /* pi/2 */
#define M_PI_4dl	0.7853981633974483096156608458198757DL  /* pi/4 */
#define M_1_PIdl	0.3183098861837906715377675267450287DL  /* 1/pi */
#define M_2_PIdl	0.6366197723675813430755350534900574DL  /* 2/pi */
#define M_2_SQRTPIdl	1.1283791670955125738961589031215452DL  /* 2/sqrt(pi) */
#define M_SQRT2dl	1.4142135623730950488016887242096981DL  /* sqrt(2) */
#define M_SQRT1_2dl	0.7071067811865475244008443621048490DL  /* 1/sqrt(2) */

#define __DECL_D(f, a, attr) __DECL_D_X(__DECIMAL, f, a, attr, __DECIMALN, )
#define __DECL_D_R(f, a, attr, r) __DECL_D_X(__DECIMAL, f, a, attr, __DECIMALN, r)
#define __DECL_D_O(rtn, f, a, attr) __DECL_D_X(rtn, f, a, attr, __DECIMALN, )
#define __DECL_D_X(rtn, f, a, attr, n, r) __DECL_D_X2(rtn, f, a, attr, n, r)
#define __DECL_D_X2(rtn, f, a, attr, n, r) \
	extern rtn f ## d ## n ## r a attr; \
	extern rtn __ ## f ## d ## n ## r a attr;

#define __DECL_D_CMP(f,a) __DECL_D_CMP_X(f,a,__DECIMALN)
#define __DECL_D_CMP_X(f,a,n) __DECL_D_CMP_X2(f,a,n)
#define __DECL_D_CMP_X2(f,a,n) extern int f ## d ## n a __THROW __attribute_const__;

#ifndef __attribute_const__
#define __attribute__ ((__const__))
#endif

__BEGIN_DECLS

#define __DECIMAL _Decimal32
#define __DECIMALN 32
#include <bits/dfp_dmathcalls.h>
#undef __DECIMAL
#undef __DECIMALN

#define __DECIMAL _Decimal64
#define __DECIMALN 64
#include <bits/dfp_dmathcalls.h>
#undef __DECIMAL
#undef __DECIMALN

#define __DECIMAL _Decimal128
#define __DECIMALN 128
#include <bits/dfp_dmathcalls.h>
#undef __DECIMAL
#undef __DECIMALN

__END_DECLS

#undef __DECL_D
#undef __DECL_D_R
#undef __DECL_D_O
#undef __DECL_D_X
#undef __DECL_D_X2
#undef __DECL_D_CMP
#undef __DECL_D_CMP_X
#undef __DECL_D_CMP_X2

/* This method works for both the GCC and the XLC compilers for detecting
 * whether a value is a Decimal Floating Point type.  */
#define __dfp_compatible(X) ((_Decimal128)(__typeof__(X))1.E-50DL == 1.E-50DL)

/* Override math.h classifications with our own. */
#undef fpclassify
#undef signbit
#undef isfinite
#undef isnormal
#undef isinf
#undef fabs
#undef isnan

# ifdef __NO_LONG_DOUBLE_MATH
#  define __dfp_ldbl(f,x) f(x)
# else
#  define __dfp_ldbl(f,x) (sizeof(x) == sizeof(double) ? f(x) : f ## l (x))
# endif

# define __dfp_dN(f,x) (sizeof (x) == sizeof(_Decimal128) ? f ## d128 (x) :   \
	(sizeof (x) == sizeof(_Decimal64) ? f ## d64 (x) : f ## d32 (x)))

/* Return number of classification appropriate for X.  */
#define __dfp_fpclassify(x) \
     (sizeof (x) == sizeof (float)					      \
      ? __fpclassifyf (x)						      \
      : __dfp_ldbl (__fpclassify, x))

#define __dfp_signbit(x) \
     (sizeof (x) == sizeof (float)					      \
      ? __signbitf (x)							      \
      : __dfp_ldbl(__signbit, x))

#define __dfp_isfinite(x) \
     (sizeof (x) == sizeof (float)					      \
      ? __finitef (x)							      \
      : __dfp_ldbl(__finite, x))

#define __dfp_isinf(x) \
     (sizeof (x) == sizeof (float)					      \
      ? __isinff (x)							      \
      : __dfp_ldbl(__isinf, x))

#define __dfp_isnan(x) \
     (sizeof (x) == sizeof (float)		      			      \
      ? __isnanf (x)				      			      \
      : __dfp_ldbl(__isnan, x))

/* A special case here.  Older glibc may not support this.  */
#ifdef issignaling
#undef issignaling
#define __dfp_issignaling(x)						      \
     (sizeof (x) == sizeof (float)					      \
      ? __issignalingf (x)						      \
      : __dfp_ldbl(__issignaling, x))
#else
#define __dfp_issignaling(x) x
#endif

#define fpclassify(x) \
  (!__dfp_compatible(x) ? __dfp_fpclassify(x) : __dfp_dN(__fpclassify, x))

#define signbit(x) \
  (!__dfp_compatible(x) ? __dfp_signbit(x) : __dfp_dN(__signbit, x))

#define isfinite(x)   \
  (!__dfp_compatible(x) ? __dfp_isfinite(x) : __dfp_dN(__isfinite, x))

#define isnormal(x) (fpclassify (x) == FP_NORMAL)

#define isinf(x) \
  (!__dfp_compatible(x)? __dfp_isinf(x) : __dfp_dN(__isinf, x))

#define isnan(x) \
  (!__dfp_compatible(x) ? __dfp_isnan(x) : __dfp_dN(__isnan, x))

#define issignaling(x) \
  (!__dfp_compatible(x) ? __dfp_issignaling(x) : __dfp_dN(__issignaling, x))

#endif /* __STDC_WANT_DEC_FP__  */

#endif /* _DFP_MATH_H  */
